### ISSUE TITLE
error: cannot find type `T` in this scope (return type of method)

### ISSUE THREAD

**Issue by Boscop** [xfix/extension-trait#1]
```rust
extension_trait! { <T> pub VecFilter for Vec<Option<T>> {
	fn filter(self) -> Vec<T> {
		self.into_iter().filter_map(|x| x).collect()
	}
}}
```
```
error[E0412]: cannot find type `T` in this scope
    |
127 |  fn filter(self) -> Vec<T> {
    |                         ^ not found in this scope
```

**Comment by sugar700** [comment]
A `VecFilter` trait needs to know what `filter` returns in its definition (the trait itself doesn't know what `T` is, only its implementation does in this example). For example, to fix this issue, `<T>` needs to be added after `VecFilter`.

```rust
extension_trait! { <T> VecFilter<T> for Vec<Option<T>> {
	fn filter(self) -> Vec<T> {
		self.into_iter().filter_map(|x| x).collect()
	}
}}

fn main() {
    let a = vec![Some(1)];
    println!("{:?}", a.filter());
}
```

Being able to use `type` inside of trait instead of trait type parameter is not implemented as of now.

It may help to think that this generates a trait, and in the following example, `T` is not defined.

```rust
trait VecFilter {
    fn filter(self) -> Vec<T>;
}
```

Probably needs to be clarified in documentation however... Or perhaps it would be feasible to simplify the syntax somehow, although I do like how it looks a lot like `impl`.

```rust
impl<T> VecFilter<T> for Vec<Option<T>> { ... }
```

**Comment by Boscop** [comment]
Ah, thanks :)

**Comment by Boscop** [comment]
And how can I add `where Self: Sized` to the trait? Or constraints on T?

**Comment by sugar700** [comment]
@Boscop 

It's possible to use `where` keyword (although this only puts a bound on an implementation, if you need a bound directly on a trait, just use `trait` directly, this crate is opinionated as a bound directly on an extension trait is VERY rarely needed).

```rust
extension_trait! { Example for str where Self: Sized {
    fn x(&self) {}
} }
```

Or just like with `impl`, put bounds directly on type variable.

```rust
extension_trait! { <T: Ord> ReverseSort for [T] {
    fn reverse_sort(&mut self) {
        self.sort_by(|a, b| b.cmp(a));
    }
} }
```

**Comment by Boscop** [comment]
So it's not possible to derive the extension trait from another trait like `trait Foo: Bar {..}` ?

**Comment by sugar700** [comment]
Not really, and I don't think it's particularly important (although usecases could be an interesting thing to hear about).

**Comment by Boscop** [comment]
Thanks. And why doesn't this work?
```rust
extension_trait! { pub PlaceExt for Place {
	/*fn location(&self) -> Location {
		Location { latitude: self.latitude, longitude: self.longitude }
	}*/
	location!();
}}
macro_rules! location {
	() => (
		fn location(&self) -> Location {
			Location { latitude: self.latitude, longitude: self.longitude }
		}
	)
}
```
> error: src\models\mod.rs:204: no rules expected the token `finish_parsing`

It works when I write it manually but I want to use the macro.
Shouldn't the inner macro be expanded before the outer macro starts parsing?

**Comment by sugar700** [comment]
@Boscop

This is a limitation of a macro system, my macro just sees tokens `location`, `!`, pair of parens with nothing in them and `;`, with no way to expand that. And even if I could expand this somehow, a macro can only return a single item/expression (which you can just use directly as is) as opposed to a token tree, which means I wouldn't be able to parse the output of a macro to replace a code block with a semicolon in trait declaration.

Technically those limitation could be dealt with by using [continuation-passing style](https://en.wikipedia.org/wiki/Continuation-passing_style), but that requires passing a macro that takes a callback. This adds complexity even on user side, as macro has to be defined in a very specific way, like this. Also this macro would have to be modified to support this style, which considering declaration wouldn't match usage I'm not particularly interesting in doing.

```rust
macro_rules! location {
    ($name:ident!($($arg:tt)*)) => (
        $name!(
            $($arg)*
            fn location(&self) -> Location {
                Location { latitude: self.latitude, longitude: self.longitude }
            }
        );
    )
}
```

**Comment by Boscop** [comment]
Nom deals with this by matching on macro calls, and then calls the macros inside the macro.
E.g. `$submac:ident!(..` here: http://rust.unhandledexpression.com/nom/macro.map_res.html
E.g. when I do `ean: map_res!(take_s!(13), |x: &str| x.parse())`, `map_res` calls `take_s!(13)` in its body.

Can't the `extension_trait` macro do the same? (sub-macro aware parsing)

**Comment by sugar700** [comment]
Nom simply includes macro call in macro output - it's not expanded by macro itself, rather Rust compiler expands macro call that was returned by a macro after macro is exited. This is very different, as it doesn't involve parsing macro output which I don't think is possible in Rust. This isn't feasible to do for `extension-trait` crate as it needs to replace a block with a semicolon in a function for a trait (trait shouldn't include code block).

I may as well write a blog article in a future about this, because it may be a long discussion topic.

**Comment by Boscop** [comment]
I'd be interested in reading that blog article :)

**Comment by Boscop** [comment]
@xfix 
> It's possible to use where keyword (although this only puts a bound on an implementation, if you need a bound directly on a trait, just use trait directly, this crate is opinionated as a bound directly on an extension trait is VERY rarely needed).

Btw, I just encountered a case where I **DID** have to use a bound directly on a trait:
```rust
extension_trait! {
    <T: TypeName + Serialize> pub Persistable for T where Self: Send {
        // ...
    }
}
```
Then I wanted to send a `Box<Persistable>` over a channel between threads, but it complained:
> the trait bound `devices::Persistable + 'static: std::marker::Send` is not satisfied

I had to rewrite it manually to place the bound directly on the trait, so I couldn't use the fancy extension_trait macro for this :(

Is there any chance that you could add to this macro the ability to place the bound on the trait directly :)
